### PR TITLE
Ruby: flow steps for ActionController filters

### PR DESCRIPTION
**Pull request by hmac**
This is the second half of https://github.com/github/codeql/pull/11666. We add
jump steps between filters in order to model the flow of execution.

### CODE REVIEW

**Review by hvitved — COMMENTED** [review]
I think it would be better to model the jump-steps as if we had synthesized the call chain. That is, if we had synthesized the calls
```rb
self.a
self.b
self.c
```
then there would be flow from _any_ post-update node inside `a` to the post-update node of `self` in the call `self.a`, and from that post-update node there would be use-use flow to `self` in `self.b`.

Now, since these calls do not actually exist, we will instead add a jump-step from any post-update node inside `a` to the `self` parameter of `b`. And to account for the use-use step from `self` in `self.b` to `self` in `self.c`, we will additionally add a jump step from the `self` parameter of `b` to the `self` parameter of `c`.

Doing it like this will have the same advantages (and drawbacks) as when we do normal flow out through arguments:

#### Drawback (1)
```rb
def a
  @foo = taint
end

def b
  @foo = safe
end

def c
  sink(@foo) # spurious flow from `taint`
end
```

#### Drawback (2)
```rb
def a
  @foo = taint
  @foo = safe
end

def b
  sink(@foo) # spurious flow from `taint`
end
```

#### Advantage (1)
```rb
def a
  @foo.bar = taint
end

def b
  sink(@foo.bar) # flow will be detected
end
```

#### Advantage (2)
```rb
def taint_foo
  @foo = taint
end

def a
  self.taint_foo
end

def b
  sink(@foo.bar) # flow will be detected
end
```

The drawbacks are well known, but historically not something that has given rise to false positive reports in practice (still, it is something we could potentially improve, at least the second drawback example).

File: ruby/ql/lib/codeql/ruby/frameworks/actioncontroller/Filters.qll [hunk 1]
```diff
@@ -387,4 +393,86 @@ module Filters {
...
+  class FilterJumpStep extends DataFlowPrivate::AdditionalJumpStep {
+    /**
+     * Holds if data can flow from `pred` to `succ` via a callback chain.
+     * `pred` is the post-update node of the self parameter in a method, and
+     * `succ` is the self parameter of a subsequent method that is executed as
+     * part of the callback chain.
+     */
+    override predicate step(DataFlow::Node pred, DataFlow::Node succ) {
+      exists(Method predMethod, Method succMethod |
+        next(predMethod, succMethod) and
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/frameworks/actioncontroller/Filters.qll:448]: nit: replace `and` with `|`, so you can get rid of the `(` indentation below.

File: ruby/ql/lib/codeql/ruby/frameworks/actioncontroller/Filters.qll [hunk 2]
```diff
@@ -387,4 +393,86 @@ module Filters {
...
+   * Holds if `n` is the syntactically last dataflow node in `m` that satisfies `selfPostUpdate`.
+   * In the example below, `n` is the post-update node for `bar`.
+   * ```rb
+   * def m
+   *   foo
+   *   bar
+   * end
+   * ```
+   */
+  predicate lastSelfPostUpdate(DataFlow::PostUpdateNode n, Method m) {
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/frameworks/actioncontroller/Filters.qll:421]: remove?

File: ruby/ql/lib/codeql/ruby/frameworks/actioncontroller/Filters.qll [hunk 3]
```diff
@@ -387,4 +393,86 @@ module Filters {
    * `pred` and `succ` may be methods bound to callbacks or controller actions.
    */
   predicate next(Method pred, Method succ) { next(_, pred, succ) }
+
+  /**
+   * Holds if `n` is a post-update node for `self` in method `m`.
+   */
+  predicate selfPostUpdate(DataFlow::PostUpdateNode n, Method m) {
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/frameworks/actioncontroller/Filters.qll:400]: private?

File: ruby/ql/lib/codeql/ruby/frameworks/actioncontroller/Filters.qll [hunk 4]
```diff
@@ -387,4 +393,86 @@ module Filters {
...
+    n.getPreUpdateNode().asExpr() instanceof SelfVariableAccessCfgNode and
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/frameworks/actioncontroller/Filters.qll:401]: This line is redundant

**Review by hvitved — COMMENTED** [review]
A few final questions, otherwise LGTM.

**Review by hvitved — DISMISSED** [review]
LGTM, but DCA reports a significant slow down. Hopefully this is just a bad join somewhere?

**Comment by hvitved** [comment]
I have reverted the last commit, and pushed another fix, which removes the following bad join:

```
[2023-02-20 11:45:21] Evaluated non-recursive predicate _AST#a6718388::AstNode::getEnclosingModule#0#dispred#ff_10#join_rhs_Call#ee92d596::MethodCallImpl#cl__#shared#1@16e09f5p in 58177ms (size: 216599109).
Evaluated relational algebra for predicate _AST#a6718388::AstNode::getEnclosingModule#0#dispred#ff_10#join_rhs_Call#ee92d596::MethodCallImpl#cl__#shared#1@16e09f5p with tuple counts:
              260   ~1%    {2} r1 = JOIN _Module#2a43f566::ModuleBase::getAMethod#0#dispred#ff_10#join_rhs_project#ActionController#32b59475:__#shared WITH Module#2a43f566::ModuleBase::getModule#0#dispred#ff ON FIRST 1 OUTPUT Rhs.1, Lhs.1
                       
            11302   ~0%    {2} r2 = JOIN r1 WITH boundedFastTC:DataFlowPublic#e1781e31::ModuleNode::getAnImmediateDescendent#0#dispred#ff_10#higher_order_body:_Module#2a43f566::ModuleBase::getAMethod#0#dispred#ff_10#join_rhs_Module#2a43f566::ModuleBase::getMo__#higher_order_body ON FIRST 1 OUTPUT Rhs.1, Lhs.1
                       
            11562   ~0%    {2} r3 = r1 UNION r2
          1870822   ~2%    {2} r4 = JOIN r3 WITH Module#2a43f566::ModuleBase::getModule#0#dispred#ff_10#join_rhs ON FIRST 1 OUTPUT Rhs.1, Lhs.1
        849865596   ~0%    {2} r5 = JOIN r4 WITH AST#a6718388::AstNode::getEnclosingModule#0#dispred#ff_10#join_rhs ON FIRST 1 OUTPUT Rhs.1, Lhs.1
        219736472   ~3%    {2} r6 = JOIN r5 WITH Call#ee92d596::MethodCallImpl#class#f ON FIRST 1 OUTPUT Lhs.0, Lhs.1
        216599109   ~2%    {2} r7 = JOIN r6 WITH CfgNodes#ace8e412::ExprCfgNode::getExpr#0#dispred#ff_10#join_rhs ON FIRST 1 OUTPUT Rhs.1, Lhs.1
                           return r7
```

**Comment by jmthomas** [comment]
Love the work here and on #11934. Question: will this enable CodeQL to track whether the various [filters](https://guides.rubyonrails.org/action_controller_overview.html#filters) halt the request cycle? Because that becomes very important if you're trying to create a filter to validate parameters.

I have a specific example of CodeQL finding "Uncontrolled data used in path expression" due to the usage of one of my parameters. I added a `before_action :sanitize_scope` and implemented the following:

```
  def sanitize_scope
    # This method is taken directly from the Rails source:
    #   https://api.rubyonrails.org/v5.2/classes/ActiveStorage/Filename.html#method-i-sanitized
    scope = params[:scope].encode(Encoding::UTF_8, invalid: :replace, undef: :replace, replace: "�").strip.tr("\u{202E}%$|:;/\t\r\n\\", "-")
    if scope != params[:scope]
      render(json: { status: 'error', message: "Invalid scope: #{params[:scope]}" }, status: 400)
    end
  end
```

This wasn't detected by CodeQL and I continue to get scan errors. I'm hoping this change will allow CodeQL to see that I'm halting the request cycle when I sanitize my input. Thoughts?

**Review by hvitved — DISMISSED** [review]
Just needs a final DCA run.

**Comment by hvitved** [comment]
> @hvitved the latest DCA still shows a slight slowdown on `canvas-lms`. Do you think that's enough to warrant looking into, or acceptable?

I think the slowdown is acceptable. I tried running the code injection query on that project, and it did reveal a bad join (but nothing too significant), which I have fixed. I'll start a final DCA run.

**Review by hvitved — APPROVED** [review]

**Pull request merged**